ops/main.py

```diff
@@ -369,97 +337,282 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
-    # For actions, there is a communication channel with the user running the
-    # action, so we want to send exception details through stderr, rather than
-    # only to juju-log as normal.
-    handling_action = ('JUJU_ACTION_NAME' in os.environ)
-    setup_root_logging(model_backend, debug=debug, exc_stderr=handling_action)
-    logger.debug("ops %s up and running.", ops.__version__)  # type:ignore
-
-    dispatcher = _Dispatcher(charm_dir)
-    dispatcher.run_any_legacy_hook()
-
-    metadata = (charm_dir / 'metadata.yaml').read_text()
-    actions_meta = charm_dir / 'actions.yaml'
-    actions_metadata = actions_meta.read_text() if actions_meta.exists() else None
-
-    # If we are in a RelationBroken event, we want to know which relation is
-    # broken within the model, not only in the event's `.relation` attribute.
-    if os.environ.get('JUJU_DISPATCH_PATH', '').endswith('-relation-broken'):
-        broken_relation_id = _get_juju_relation_id()
-    else:
-        broken_relation_id = None
-    meta = CharmMeta.from_yaml(metadata, actions_metadata)
-    model = ops.model.Model(meta, model_backend, broken_relation_id=broken_relation_id)
-
-    charm_state_path = charm_dir / CHARM_STATE_FILE
-
-    if use_juju_for_storage and not ops.storage.juju_backend_available():
-        # raise an exception; the charm is broken and needs fixing.
-        msg = 'charm set use_juju_for_storage=True, but Juju version {} does not support it'
-        raise RuntimeError(msg.format(JujuVersion.from_environ()))
-
-    if use_juju_for_storage is None:
-        use_juju_for_storage = _should_use_controller_storage(charm_state_path, meta)
-    elif use_juju_for_storage:
-        warnings.warn("Controller storage is deprecated; it's intended for "
-                      "podspec charms and will be removed in a future release.",
-                      category=DeprecationWarning)
-
-    if use_juju_for_storage:
-        if dispatcher.is_restricted_context():
+    CHARM_STATE_FILE = '.unit-state.db'
+
+    def __init__(
+            self,
+            charm_spec: "_CharmSpec",
+            use_juju_for_storage: Optional[bool] = None
+    ):
+        self.charm_spec = charm_spec
+        self._charm_meta = CharmMeta(charm_spec.meta, charm_spec.actions)
+        self._use_juju_for_storage = use_juju_for_storage
+
+        # set by setup()
+        self.dispatcher: Optional[_Dispatcher] = None
+        self.framework: Optional["ops.framework.Framework"] = None
+        self.charm: Optional["ops.charm.CharmBase"] = None
+
+        self._has_setup = False
+        self._has_emitted = False
+        self._has_committed = False
+
+    def _setup_charm(self, framework: "ops.framework.Framework", dispatcher: _Dispatcher):
+        charm_type = self.charm_spec.charm_type
+
+        charm = charm_type(framework)
+        dispatcher.ensure_event_links(charm)
+        return charm
+
+    @staticmethod
+    def _setup_model_backend():
```

Comment:
Would it be simpler just to have a `backend_class=ops.model._ModelBackend` parameter to `_Ops.__init__`?